### PR TITLE
DTSRD-2139.Update Organisation user idam id endpoint

### PR DESCRIPTION
**Pull request by SabinaHMCTS**
### Jira link (if applicable)

https://tools.hmcts.net/jira/browse/DTSRD-2139

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        userProfileCreateUserWireMock(HttpStatus.CREATED);
+        updateUserProfileRolesMock(HttpStatus.BAD_REQUEST);
+
+        Map<String, Object> modifiedUserResponse =
+            professionalReferenceDataClient.updateUserIdamForOrganisation(professionalUserIdentifierRequest,hmctsAdmin);
+
+        assertThat(modifiedUserResponse.get("http_status")).isNotNull();
+        assertThat(modifiedUserResponse).containsEntry("http_status","400");
+    }
+}
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java:109]: add error scenario test cases for below user profile calls:
1. userProfileFeignClient.getUserProfileById 
2. userProfileFeignClient.modifyUserRoles. (change this method name as its confusing)

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+        userProfileCreateUserWireMock(HttpStatus.CREATED);
+        updateUserProfileRolesMock(HttpStatus.BAD_REQUEST);
+
+        Map<String, Object> modifiedUserResponse =
+            professionalReferenceDataClient.updateUserIdamForOrganisation(professionalUserIdentifierRequest,hmctsAdmin);
+
+        assertThat(modifiedUserResponse.get("http_status")).isNotNull();
+        assertThat(modifiedUserResponse).containsEntry("http_status","400");
+    }
```
Comment (rajkatla-hmcts) [inline src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java:108]: add test case for error scenario:  "Request is empty"

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java [hunk 3]
```diff
@@ -689,6 +691,48 @@ void findOrganisationByPbaStatusShouldReturn403WhenToggledOff() {
...
+        OrganisationCreationRequest orgReq = createOrganisationRequest().build();
+        Map<String, Object> response = professionalApiClient.createOrganisation(orgReq);
+        String orgIdentifier = (String) response.get("organisationIdentifier");
+        orgReq.setStatus("ACTIVE");
+        professionalApiClient.updateOrganisation(orgReq, hmctsAdmin,
+            orgIdentifier, OK);
+
+        NewUserCreationRequest newUserCreationRequest =
+            professionalApiClient.createNewUserRequest("testEmail@hmcts.com");
+        Map<String, Object> newUserResponse = professionalApiClient.addNewUserToAnOrganisation(orgIdentifier,
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java:709]: creating a new user as test user to use as existing user for updating id

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java [hunk 4]
```diff
@@ -689,6 +691,48 @@ void findOrganisationByPbaStatusShouldReturn403WhenToggledOff() {
...
+        Map<String, Object> updatedIdamResponse = professionalApiClient.updateUserIdamForOrganisation(
+            professionalUserIdentifierRequest, OK);
+
+        Map<String, Object> searchResponse = professionalApiClient
+            .searchOrganisationUsersByStatusInternal(orgIdentifier, hmctsAdmin, OK);
+
+        List<Map<String, Object>> professionalUsersResponses = (List<Map<String, Object>>) searchResponse.get("users");
+        Map professionalUsersResponse = getUserById(professionalUsersResponses, existingUserId);
+
+        professionalApiClient.deleteOrganisation(orgIdentifier, hmctsAdmin, NO_CONTENT);
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java:728]: done

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Test
+    void ac2_modify_idam_newId_null_should_return_400() {
+
+        //create and update organisation
+        String organisationIdentifier = createOrganisationRequest();
+        updateOrganisation(organisationIdentifier, hmctsAdmin, ACTIVE);
+
+        //create new user
+        List<String> userRoles = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java:67]: creatign test user

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+class ModifyUserIdamIdIntegrationTest extends AuthorizationEnabledIntegrationTest {
+
+    @Test
+    void ac1_modify_idamId_of_active_user_for_an_active_organisation_should_return_200() {
+        //create and update organisation
+        String organisationIdentifier = createOrganisationRequest();
+        updateOrganisation(organisationIdentifier, hmctsAdmin, ACTIVE);
+
+        //create new user
+        List<String> userRoles = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java:29]: creating test user

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java (hunk 2 above)
Comment (SabinaHMCTS) [reply src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java:108]: done

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java [hunk 7]
```diff
@@ -697,6 +700,73 @@ public ResponseEntity<OrganisationEntityResponse> retrieveOrganisationByUserId(
         return organisationService.retrieveOrganisationByUserId(userId);
     }
 
+
+    @Operation(
+        summary = "Updates idam id for a user in an organisation",
+        description = "**IDAM Roles to access API** :<br> pui-organisation-manager",
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java:706]: done

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        userProfileCreateUserWireMock(HttpStatus.CREATED);
+
+        updateUserProfileRolesMock(HttpStatus.OK);
+
+        Map<String, Object> modifiedUserResponse =
+            professionalReferenceDataClient.updateUserIdamForOrganisation(professionalUserIdentifierRequest,hmctsAdmin);
+
+
+        //validate overall response should be 200 always
+        assertThat(modifiedUserResponse.get("http_status")).isNotNull();
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java:53]: not fetching id in response

File: src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/OrganisationServiceImpl.java [hunk 9]
```diff
@@ -1056,5 +1067,50 @@ private boolean getMoreAvailable(Page<Organisation> pageableOrganisations) {
...
+            GetUserProfileResponse existingUserProfile =
+                (GetUserProfileResponse) requireNonNull(responseResponseEntity.getBody());
+
+            UserProfileUpdatedData userProfileUpdatedData = UserProfileUpdatedData.builder()
+                .idamId(newUserIdentifier)
+                .idamStatus(existingUserProfile.getIdamStatus().name())
+                .build();
+
+            Response userResponse = userProfileFeignClient
+                .modifyUserRoles(userProfileUpdatedData, existingUserIdentifier," ");
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/OrganisationServiceImpl.java:1093]: ### using existing method cannot change name

File: src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/OrganisationServiceImpl.java [hunk 10]
```diff
@@ -1056,5 +1067,50 @@ private boolean getMoreAvailable(Page<Organisation> pageableOrganisations) {
...
+
+            Optional<ProfessionalUser> professionalUser = Optional
```
[the remark this answers is not in the document]
Comment (SabinaHMCTS) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/OrganisationServiceImpl.java:1095]: done

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java (hunk 1 above)
Comment (SabinaHMCTS) [reply src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/ModifyUserIdamIdIntegrationTest.java:125]: added error method
